diff --git a/bfd/elfxx-riscv.c b/bfd/elfxx-riscv.c
index c0f006f5e5..aa9271444a 100644
--- a/bfd/elfxx-riscv.c
+++ b/bfd/elfxx-riscv.c
@@ -855,6 +855,39 @@ static reloc_howto_type howto_table[] =
 	 0,				/* src_mask */
 	 MINUS_ONE,			/* dst_mask */
 	 FALSE),			/* pcrel_offset */
+
+	//JMNOTE:UVEBRANCH
+	/* 12-bit PC-relative branch offset.  */
+   HOWTO (R_RISCV_NULL_PLACE,		/* type */
+	 0,				/* rightshift */
+	 2,				/* size */
+	 32,				/* bitsize */
+	 TRUE,				/* pc_relative */
+	 0,				/* bitpos */
+	 complain_overflow_signed,	/* complain_on_overflow */
+	 bfd_elf_generic_reloc,		/* special_function */
+	 "R_RISCV_NULL_PLACE",		/* name */
+	 FALSE,				/* partial_inplace */
+	 0,				/* src_mask */
+	 ENCODE_UVEATYPE_IMM (-1U),	/* dst_mask */
+	 TRUE),				/* pcrel_offset */
+	
+    //JMNOTE:UVEBRANCH
+	/* 12-bit PC-relative branch offset.  */
+   HOWTO (R_RISCV_UVE_STREAM_BRANCH,		/* type */
+	 0,				/* rightshift */
+	 2,				/* size */
+	 32,				/* bitsize */
+	 TRUE,				/* pc_relative */
+	 0,				/* bitpos */
+	 complain_overflow_signed,	/* complain_on_overflow */
+	 bfd_elf_generic_reloc,		/* special_function */
+	 "R_RISCV_UVE_STREAM_BRANCH",		/* name */
+	 FALSE,				/* partial_inplace */
+	 0,				/* src_mask */
+	 ENCODE_UVEATYPE_IMM (-1U),	/* dst_mask */
+	 TRUE),				/* pcrel_offset */
+
 };
 
 /* A mapping from BFD reloc types to RISC-V ELF reloc types.  */
@@ -917,6 +950,8 @@ static const struct elf_reloc_map riscv_reloc_map[] =
   { BFD_RELOC_RISCV_SET16, R_RISCV_SET16 },
   { BFD_RELOC_RISCV_SET32, R_RISCV_SET32 },
   { BFD_RELOC_RISCV_32_PCREL, R_RISCV_32_PCREL },
+  { BFD_RELOC_RISCV_NULL_PLACE, R_RISCV_NULL_PLACE },
+  { BFD_RELOC_RISCV_UVE_STREAM_BRANCH, R_RISCV_UVE_STREAM_BRANCH },
 };
 
 /* Given a BFD reloc type, return a howto structure.  */
